table/tables/tables_test.go

```diff
@@ -943,3 +943,116 @@ func TestTxnAssertion(t *testing.T) {
+	if *testkit.WithTiKV == "" {
+		tk.MustExec("set @@tidb_txn_assertion_level=off")
+	}
+
+	// case: success
+	tk.MustExec("begin pessimistic")
+	tk.MustExec("insert into t values (1, 1)")
+	tk.MustQuery("select * from t for update").Check(testkit.Rows("1 1"))
+	tk.MustExec("commit")
+	tk.MustExec("admin check table t")
+
+	tk.MustExec("truncate table t")
+	tk2.MustExec("insert into t values (2, 1)")
+	tk.MustExec("begin pessimistic")
+	tk.MustExec("insert into t values (2, 2)")
+	tk2.MustExec("delete from t")
+	tk.MustQuery("select * from t for update").Check(testkit.Rows("2 2"))
+	tk.MustExec("commit")
+	tk.MustExec("admin check table t")
+
+	// case: constraint check failure
+	tk.MustExec("create table t2 (id int primary key, uk int, unique key i1(uk))")
+	tk.MustExec("insert into t2 values (1, 1)")
+	tk.MustExec("begin pessimistic")
+	tk.MustExec("insert into t2 values (2, 1), (3, 3)")
+	// NOTE: this read breaks constraint, but we are not able to return an error here.
+	// We can only guarantee the txn should not commit
+	tk.MustQuery("select * from t2 use index(primary) for update").Check(testkit.Rows("1 1", "2 1", "3 3"))
+	err := tk.ExecToErr("commit")
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "Duplicate entry '1' for key 'i1'")
+	tk.MustQuery("select * from t2 use index(primary)").Check(testkit.Rows("1 1"))
+	tk.MustExec("admin check table t2")
+
+	// case: conflict check failure
+	tk.MustExec("create table t3 (id int primary key, sk int, key i1(sk))")
+	tk.MustExec("begin pessimistic")
+	tk.MustExec("insert into t3 values (1, 1)")
+	tk2.MustExec("insert into t3 values (1, 2)")
+	err = tk.ExecToErr("commit")
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "[kv:9007]Write conflict")
+
+	// case: DML returns error => abort txn
+	tk.MustExec("create table t4 (id int primary key, v int, key i1(v))")
+	tk.MustExec("insert into t4 values (1, 1)")
+	tk.MustExec("begin pessimistic")
+	tk.MustExec("insert into t4 values (1, 2), (2, 2)")
+	tk.MustQuery("select * from t4 order by id").Check(testkit.Rows("1 2", "2 2"))
+	err = tk.ExecToErr("delete from t4 where id = 1")
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "transaction aborted because lazy uniqueness check is enabled and an error occurred: [kv:1062]Duplicate entry '1' for key 'PRIMARY'")
+	tk.MustExec("commit")
+	tk.MustExec("admin check table t4")
+	tk.MustQuery("select * from t4 order by id").Check(testkit.Rows("1 1"))
+
+	// case: larger for_update_ts should not prevent the "write conflict" error.
+	tk.MustExec("create table t5 (id int primary key, uk int, unique key i1(uk))")
+	tk.MustExec("insert into t5 values (1, 1), (2, 2)")
+	tk.MustExec("begin pessimistic")
+	tk.MustExec("update t5 set uk = 2 where id = 1")
+	tk2.MustExec("delete from t5 where uk = 2")
+	tk.MustExec("select * from t5 for update")
+	err = tk.ExecToErr("commit")
+	require.Contains(t, err.Error(), "[kv:9007]Write conflict")
+}
```

Comment:
How about adding cases for `update` statements that the newly inserted value cause conflicts but the check is skipped.

Reply:
Also other extended statements like `insert ignore` or `insert on duplicate update` should not be affected if it's enabled, maybe we should add cases to verify this?

Reply:
`Insert on duplicate update` is similar to an update when there are duplicates, and should be affected, I suppose.

Reply:
If it's supposed to be supported related tests need to be added.

Reply:
Yeah I've added them with cases for `update`.